### PR TITLE
Travis stages

### PR DESCRIPTION
**Pull request by kevnm67**
## Pull request checklist

- [x] All tests pass. 
- [x] Demo project builds and runs.
- [x] I have resolved merge conflicts.
- [x] I have followed the [coding style](https://github.com/jessesquires/HowToContribute#style-guidelines) and [contributing guidelines](https://github.com/jessesquires/HowToContribute).

#### This fixes issue #
- #64 and builds on #129

## What's in this pull request?
- Updated travis to use build stages vs. matrix.
- [Added] Build carthage stage.

### CODE REVIEW

**Comment by kevnm67** [comment]
@jessesquires Got this [working](https://github.com/kevnm67/JSQDataSourcesKit/pull/4) over the weekend.  I added #64.  Happy to break it out into a follow-up PR, though.

Let me know what you think and want any changes.

File: .travis.yml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 os: osx
 osx_image: xcode10.1
+language: swift
```
Comment (jessesquires) [inline .travis.yml]: do we need this? 

i think only specifying `os:` and `osx_image:` should work, right?

File: .travis.yml [hunk 2]
```diff
@@ -23,35 +24,58 @@ env:
...
   - gem install bundler
 
-script:
-- set -o pipefail
-- xcodebuild -version -sdk
+jobs:
+  include:
+    - &test
+      stage: test
+      name: iOS iPhone X 11.4
```
Comment (jessesquires) [inline .travis.yml]: can we just use the OS name? 

so: `iOS 11.4`, `iOS 12.1`, `tvOS 12.1`

File: .travis.yml [hunk 3]
```diff
@@ -23,35 +24,58 @@ env:
...
+      env: DESTINATION="OS=11.4,name=iPhone X" SDK="$IOS_SDK" BUILD_EXAMPLE="YES"
```
Comment (jessesquires) [inline .travis.yml]: we can remove `BUILD_EXAMPLE="YES/NO"` from all of these 😊

File: .travis.yml [hunk 4]
```diff
@@ -23,35 +24,58 @@ env:
...
+      script:
+        - set -o pipefail
+        - xcodebuild -version -sdk
+        - xcodebuild clean test -project "$PROJECT" -scheme JSQDataSourcesKit -sdk "$SDK" -destination "$DESTINATION" -enableCodeCoverage YES -configuration Debug ONLY_ACTIVE_ARCH=NO CODE_SIGNING_REQUIRED=NO | bundle exec xcpretty -c;
```
Comment (jessesquires) [inline .travis.yml]: now that I think about it, let's pull `JSQDataSourcesKit` out into a `$SCHEME` variable.

then this travis file will be easier to reuse in other projects 😄

File: .travis.yml [hunk 5]
```diff
@@ -23,35 +24,58 @@ env:
...
+      env: DESTINATION="OS=12.1,name=iPhone X" SDK="$IOS_SDK" BUILD_EXAMPLE="YES"
 
-- xcodebuild analyze test -project "$PROJECT" -scheme JSQDataSourcesKit -sdk "$SDK" -destination "$DESTINATION" -enableCodeCoverage YES -configuration Debug ONLY_ACTIVE_ARCH=NO CODE_SIGNING_REQUIRED=NO | bundle exec xcpretty -c;
+    - <<: *test
+      name: Apple TV 12.1
+      env: DESTINATION="OS=12.1,name=Apple TV" SDK="$TVOS_SDK" BUILD_EXAMPLE="NO"
 
-- bundle exec danger
+    - &build
+      stage: build
```
Comment (jessesquires) [inline .travis.yml]: can we call this `build examples` ?

File: .travis.yml [hunk 6]
```diff
@@ -23,35 +24,58 @@ env:
...
-    script: bundle exec pod lib lint
+    - stage: lint
+      name: "pod lint"
+      script: bundle exec pod lib lint
+
+    - stage: carthage
+      osx_image: xcode10.1
+      name: Carthage
+      os: osx
+      script: carthage build --no-skip-current --configuration Release
```
Comment (jessesquires) [inline .travis.yml:75]: nice! 👏 👏 👏

File: .travis.yml [hunk 7]
```diff
@@ -23,35 +24,58 @@ env:
...
+
+stages:
+  - test
+  - build
+  - lint
+  - carthage
```
Comment (jessesquires) [inline .travis.yml:81]: 💯

**Review by jessesquires — CHANGES_REQUESTED** [review]
thanks @kevnm67 ! 

this is awesome 💯 🥇 

only a few minor comments

File: .travis.yml (hunk 1 above)
Comment (kevnm67) [reply .travis.yml]: I go back and forth on this.  Sorry about that.  Ill remove it.

image


FYI, I found the [thread](https://github.com/travis-ci/docs-travis-ci-com/issues/922) about Travis CI treating ```objective-c``` and ```swift``` the same.

File: .travis.yml (hunk 2 above)
Comment (kevnm67) [reply .travis.yml]: Sounds good.  Ill update those now.

File: .travis.yml (hunk 3 above)
Comment (kevnm67) [reply .travis.yml]: Ha, of course!

File: .travis.yml (hunk 5 above)
Comment (kevnm67) [reply .travis.yml]: Done

File: .travis.yml (hunk 6 above)
Comment (kevnm67) [reply .travis.yml:75]: :) Thought I'd throw it in here... and ask for forgiveness if you requested to break it out 😺

File: .travis.yml (hunk 7 above)
Comment (kevnm67) [reply .travis.yml:81]: One question:  
Do you want to conditionally build any stage(s)?  (i.e. only run ```carthage``` if branch == master && !PR).

Comment (jessesquires) [reply .travis.yml:81]: meh, i think it's fine to always run. it shouldn't take long

(and i'd rather catch bugs early)

File: .travis.yml (hunk 1 above)
Comment (jessesquires) [reply .travis.yml]: 💯

**Review by jessesquires — APPROVED** [review]

**Comment by jessesquires** [comment]
thanks again @kevnm67 ! 🌈 

this is fucking awesome

**Comment by kevnm67** [comment]
@jessesquires Thanks! I appreciate the kind words 😎

**Pull request merged**